File: commons/org.openmole.commons.aspect.caching/src/main/java/org/openmole/commons/aspect/caching/MethodCache.java
package org.openmole.commons.aspect.caching;

import java.util.Map;
import java.util.TreeMap;
import java.util.WeakHashMap;

public class MethodCache {

//    final static Logger LOGGER = Logger.getLogger(MethodCache.class.getName());
    
    final Map<Object, Map<String, Object>> cache = new WeakHashMap<Object, Map<String, Object>>();

    void putCachedMethodResult(Object object, String method, Object result) {
        getMethodMap(object).put(method, result);
    }

    Object getCachedMethodResult(Object object, String method) {
        Map<String, Object> methodMap = cache.get(object);
        if (methodMap == null) {
            return null;
        }
        Object toReturn = methodMap.get(method);
        
//        LOGGER.log(Level.FINE, "Cached {0} size {1} returning {2}", new Object[]{method, size(), toReturn});
 
        return toReturn;
    }

    private Map<String, Object> getMethodMap(Object object) {
        Map<String, Object> methodMap = cache.get(object);
        if (methodMap != null) {
            return methodMap;
        }

        synchronized (cache) {
            methodMap = cache.get(object);
            if (methodMap == null) {
                methodMap = new TreeMap<String, Object>();
                cache.put(object, methodMap);
            }
            return methodMap;
        }
    }

    void clear(Object obj) {
        cache.remove(obj);
    }

    public int size() {
        return cache.size();
    }
}

File: commons/org.openmole.commons.aspect.caching/src/main/scala/org/openmole/commons/aspect/caching/MethodCache.scala
/*
 * Copyright (C) 2011 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.commons.aspect.caching

import scala.collection.mutable.HashMap
import scala.collection.mutable.SynchronizedMap
import scala.collection.mutable.WeakHashMap

class MethodCache {
    
    val cache = new WeakHashMap[Object, HashMap[String, Object]] with SynchronizedMap[Object, HashMap[String, Object]]

    def putCachedMethodResult(obj: Object, method: String, result: Object) = {
        val methodMap = cache.getOrElseUpdate(obj, new HashMap[String, Object] with SynchronizedMap[String, Object])
        methodMap.put(method, result)
    }

    def cachedMethodResult(obj: Object, method: String): Object = {
        cache.get(obj) match {
          case None => return null
          case Some(methodMap) => methodMap.getOrElse(method, null)
        }

    }


    def clear(obj: Object) = cache.remove(obj)

    def size = cache.size
    
}
